### PR TITLE
Multi tenancy kubectl commands

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR adds explicit multi-tenancy support in CLI options to multiple kubectl sub-commands. With the change, the system-tenant user can use these commands to operate on resources of other tenants.

Subcommands updated:
* annotate
* label
* create job
* patch
* run
* autoscale
* diff
* cluster-ino
* expose
* drain
* cp
* taint

Details about the subcommands updated in this PR are:
## kubectl annotate
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl annotate deployment sample-nginx-deployment aaa="bbb" --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment annotated

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get deployment sample-nginx-deployment --tenant sample-tenant-1 --namespace sample-namespace-a -o yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    aaa: bbb
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx"},"name":"sample-nginx-deployment","namespace":"sample-namespace-a","tenant":"sample-tenant-1"},"spec":{"replicas":3,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx:1.7.9","name":"nginx","ports":[{"containerPort":80}]}]}}}}
 ....
```
## kubectl label
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl label deployment sample-nginx-deployment ccc="ddd" --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment labeled

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get deployment sample-nginx-deployment --tenant sample-tenant-1 --namespace sample-namespace-a -o yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
...
  labels:
    app: nginx
    ccc: ddd
  name: sample-nginx-deployment
  namespace: sample-namespace-a
  resourceVersion: "707"
  selfLink: /apis/extensions/v1beta1/tenants/sample-tenant-1/namespaces/sample-namespace-a/deployments/sample-nginx-deployment
  tenant: sample-tenant-1
  uid: 68bd1bf8-5367-48e3-bc67-6e417bc14cce
...
```
## kubectl create job
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl create job my-job --image=buysbox --tenant aaa
job.batch/my-job created
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get jobs --all-tenants --all-namespaces
TENANT   NAMESPACE   NAME     COMPLETIONS   DURATION   AGE
aaa      default     my-job   0/1           20s        20s
```
## kubectl patch
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ cat ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/patch-file.yaml
spec:
  template:
    spec:
      containers:
      - name: patch-demo-ctr-2
        image: redis

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get deployment sample-nginx-deployment  --tenant sample-tenant-1 --namespace sample-namespace-a -o yamlapiVersion: extensions/v1beta1
kind: Deployment
metadata:
  ...
  name: sample-nginx-deployment
  namespace: sample-namespace-a
  resourceVersion: "707"
  selfLink: /apis/extensions/v1beta1/tenants/sample-tenant-1/namespaces/sample-namespace-a/deployments/sample-nginx-deployment
  tenant: sample-tenant-1
  uid: 68bd1bf8-5367-48e3-bc67-6e417bc14cce
spec:
  progressDeadlineSeconds: 600
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx:1.7.9
        imagePullPolicy: IfNotPresent
        name: nginx
        ports:
        - containerPort: 80
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30


qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl patch deployment sample-nginx-deployment --patch "$(cat ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/patch-file.yaml)" --tenant sample-tenant-1 --namespace sample-namespace-a
deployment.extensions/sample-nginx-deployment patched

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get deployment sample-nginx-deployment  --tenant sample-tenant-1 --namespace sample-namespace-a -o yamlapiVersion: extensions/v1beta1
kind: Deployment
metadata:
  ...
  name: sample-nginx-deployment
  namespace: sample-namespace-a
  resourceVersion: "955"
  selfLink: /apis/extensions/v1beta1/tenants/sample-tenant-1/namespaces/sample-namespace-a/deployments/sample-nginx-deployment
  tenant: sample-tenant-1
  uid: 68bd1bf8-5367-48e3-bc67-6e417bc14cce
spec:
  progressDeadlineSeconds: 600
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: nginx
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: nginx
    spec:
      containers:
      - image: redis
        imagePullPolicy: Always
        name: patch-demo-ctr-2
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      - image: nginx:1.7.9
        imagePullPolicy: IfNotPresent
        name: nginx
        ports:
        - containerPort: 80
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
...
```
## kubectl run
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl run nginx --image=nginx --replicas=1 --generator=run-pod/v1 --tenant aaa
pod/nginx created

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get pods --tenant aaa --all-namespaces
TENANT            NAMESPACE            NAME                                       HASHKEY               READY   STATUS                 RESTARTS   AGE
aaa               default              nginx                                      5415868506610159940   1/1     Running                0          29s
```
## kubectl autoscale
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl autoscale deployment sample-nginx-deployment --cpu-percent=50 --min=1 --max=10 --tenant sample-tenant-1 --namespace sample-namespace-a
horizontalpodautoscaler.autoscaling/sample-nginx-deployment autoscaled

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get horizontalpodautoscalers --all-tenants --all-namespaces
TENANT            NAMESPACE            NAME                      REFERENCE                            TARGETS         MINPODS   MAXPODS   REPLICAS   AGE
sample-tenant-1   sample-namespace-a   sample-nginx-deployment   Deployment/sample-nginx-deployment   <unknown>/50%   1         10        0          45s
```
## kubectl diff
"--tenant" option is in effect now. 

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ cat ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/no-tenant-no-ns-a-deployment.yaml 
apiVersion: apps/v1
kind: Deployment
metadata:
  name: sample-nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80


qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl diff --tenant aaa --namespace sample-namespace-a -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/no-tenant-no-ns-a-deployment.yaml
diff -u -N /tmp/LIVE-179312307/apps.v1.Deployment.aaa.sample-namespace-a.sample-nginx-deployment /tmp/MERGED-669412470/apps.v1.Deployment.aaa.sample-namespace-a.sample-nginx-deployment
--- /tmp/LIVE-179312307/apps.v1.Deployment.aaa.sample-namespace-a.sample-nginx-deployment	2020-06-17 00:23:11.516375060 -0700
+++ /tmp/MERGED-669412470/apps.v1.Deployment.aaa.sample-namespace-a.sample-nginx-deployment	2020-06-17 00:23:11.520375087 -0700
@@ -0,0 +1,47 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  creationTimestamp: "2020-06-17T07:23:11Z"
+  generation: 1
+  hashKey: 2429510213523666571
+  labels:
+    app: nginx
+  name: sample-nginx-deployment
+  namespace: sample-namespace-a
+  selfLink: /apis/apps/v1/tenants/aaa/namespaces/sample-namespace-a/deployments/sample-nginx-deployment
+  tenant: aaa
+  uid: 94750544-33ae-4219-8929-bfe567556ca1
+spec:
+  progressDeadlineSeconds: 600
+  replicas: 3
+  revisionHistoryLimit: 10
+  selector:
+    matchLabels:
+      app: nginx
+  strategy:
+    rollingUpdate:
+      maxSurge: 25%
+      maxUnavailable: 25%
+    type: RollingUpdate
+  template:
+    metadata:
+      creationTimestamp: null
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - image: nginx:1.7.9
+        imagePullPolicy: IfNotPresent
+        name: nginx
+        ports:
+        - containerPort: 80
+          protocol: TCP
+        resources: {}
+        terminationMessagePath: /dev/termination-log
+        terminationMessagePolicy: File
+      dnsPolicy: ClusterFirst
+      restartPolicy: Always
+      schedulerName: default-scheduler
+      securityContext: {}
+      terminationGracePeriodSeconds: 30
+status: {}
exit status 1
```
## kubectl cluster-info
"--tenant" option is in effect now. This commad will show different networking config for different tenants.

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl cluster-info --tenant aaa
Kubernetes master is running at http://localhost:8080

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl cluster-info --tenant system
Kubernetes master is running at http://localhost:8080
KubeDNS is running at http://localhost:8080/api/v1/tenants/system/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
```

## kubectl expose

"--tenant" option is in effect now. Without this change, system tenant user can expose a regular-tenant resource via kubectl.

test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl expose deployment sample-nginx-deployment --port=80 --target-port=8080 --tenant sample-tenant-1 --namespace sample-namespace-a
service/sample-nginx-deployment exposed
```
## kubectl drain

Changes are made so that non-system-tenant pods can be evicted. Before this change, only system-tenant pods will be evicted.
no change to "--tenant" option this command as the command target is node, which is not tenant-scoped. 

test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl drain qianchen-virtualbox --ignore-daemonsets
node/qianchen-virtualbox cordoned
WARNING: ignoring DaemonSet-managed Pods: system/kube-system/virtlet-wlzvl
evicting pod "sample-nginx-deployment-74cbdd5d99-tvz8g"
evicting pod "kube-dns-5d69f5657d-hrr24"
evicting pod "sample-nginx-deployment-74cbdd5d99-fs752"
evicting pod "sample-nginx-deployment-74cbdd5d99-r645l"
pod/sample-nginx-deployment-74cbdd5d99-tvz8g evicted
pod/sample-nginx-deployment-74cbdd5d99-fs752 evicted
pod/sample-nginx-deployment-74cbdd5d99-r645l evicted
pod/kube-dns-5d69f5657d-hrr24 evicted
node/qianchen-virtualbox evicted
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get pods --all-tenants --all-namespaces
TENANT            NAMESPACE            NAME                                       HASHKEY               READY   STATUS    RESTARTS   AGE
system            kube-system          kube-dns-5d69f5657d-xlk6b                  1893193774514411102   0/3     Pending   0          42s
system            kube-system          virtlet-wlzvl                              2054872789227874544   3/3     Running   0          54s
sample-tenant-2   sample-namespace-a   sample-nginx-deployment-74cbdd5d99-2p2mr   3114283537275519151   0/1     Pending   0          42s
sample-tenant-2   sample-namespace-a   sample-nginx-deployment-74cbdd5d99-cnzxr   3075577187818199037   0/1     Pending   0          42s
sample-tenant-2   sample-namespace-a   sample-nginx-deployment-74cbdd5d99-jdk4p   6440251859956490430   0/1     Pending   0          42s
```
## kubectl cp
"--tenant" option is in effect. Changes are also made so copy can happen to a non-system-tenant pod via a namespaced name with tenant name included.

### test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl cp /home/qianchen/go113/go/src/archive/tar/testdata/hardlink.tar default/nginx:/tmp/ --tenant aaa

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl exec --tenant aaa -it nginx ls /tmp
hardlink.tar

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl cp /home/qianchen/go113/go/src/archive/tar/testdata/writer-big.tar aaa/default/nginx:/tmp/ 

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl exec --tenant aaa -it nginx ls /tmp
hardlink.tar  writer-big.tar
```

## kubectl tainted

Changes are made so that non-system-tenant resources can be handled. Before this change, only system-tenant resources are handled"
no change to "--tenant" option this command as the command target is node, which is not tenant-scoped. 

test done
```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get pods --tenant aaa
NAME    HASHKEY               READY   STATUS    RESTARTS   AGE
nginx   8926768411090435021   1/1     Running   0          19s

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl taint node qianchen-virtualbox aaa=bbb:NoExecute
node/qianchen-virtualbox tainted

qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl get pods --tenant aaa
No resources found.
```

### CODE REVIEW

File: pkg/controller/nodelifecycle/scheduler/timed_workers_test.go [hunk 1]
```diff
@@ -62,16 +65,16 @@ func TestExecuteDelayed(t *testing.T) {
...
-	queue.AddWork(NewWorkArgs("2", "2"), now, then)
-	queue.AddWork(NewWorkArgs("3", "3"), now, then)
-	queue.AddWork(NewWorkArgs("4", "4"), now, then)
-	queue.AddWork(NewWorkArgs("5", "5"), now, then)
-	queue.AddWork(NewWorkArgs("1", "1"), now, then)
-	queue.AddWork(NewWorkArgs("2", "2"), now, then)
-	queue.AddWork(NewWorkArgs("3", "3"), now, then)
-	queue.AddWork(NewWorkArgs("4", "4"), now, then)
-	queue.AddWork(NewWorkArgs("5", "5"), now, then)
+	queue.AddWork(NewWorkArgs("1", "1", metav1.TenantSystem), now, then)
```
Comment (pdgetrf) [inline pkg/controller/nodelifecycle/scheduler/timed_workers_test.go:68]: curious: what would happen if these were called using a regular tenant.

File: pkg/kubectl/cmd/clusterinfo/clusterinfo.go [hunk 2]
```diff
@@ -138,9 +144,9 @@ func (o *ClusterInfoOptions) Run() error {
 				}
 
 				if len(client.GroupVersion.Group) == 0 {
-					link = client.Host + "/api/" + client.GroupVersion.Version + "/namespaces/" + service.ObjectMeta.Namespace + "/services/" + name + "/proxy"
+					link = client.Host + "/api/" + client.GroupVersion.Version + "/tenants/" + service.ObjectMeta.Tenant + "/namespaces/" + service.ObjectMeta.Namespace + "/services/" + name + "/proxy"
 				} else {
-					link = client.Host + "/api/" + client.GroupVersion.Group + "/" + client.GroupVersion.Version + "/namespaces/" + service.ObjectMeta.Namespace + "/services/" + name + "/proxy"
+					link = client.Host + "/api/" + client.GroupVersion.Group + "/" + client.GroupVersion.Version + "/tenants/" + service.ObjectMeta.Tenant + "/namespaces/" + service.ObjectMeta.Namespace + "/services/" + name + "/proxy"
```
Comment (pdgetrf) [inline pkg/kubectl/cmd/clusterinfo/clusterinfo.go:149]: make these two long string concatenation into a single function.

File: pkg/kubectl/cmd/cp/cp.go [hunk 3]
```diff
@@ -137,6 +140,14 @@ func extractFileSpec(arg string) (fileSpec, error) {
 				File:         file,
 			}, nil
 		}
+		if len(pieces) == 3 {
```
Comment (pdgetrf) [inline pkg/kubectl/cmd/cp/cp.go:143]: use a constant for 3 with meaningful name or add a comment explaining why 3?

File: pkg/controller/nodelifecycle/scheduler/timed_workers_test.go (hunk 1 above)
Comment (chenqianfzh) [reply pkg/controller/nodelifecycle/scheduler/timed_workers_test.go:68]: That should work for a regular tenant. The manual test cases I gave in the PR description are for cases where the tenant are not system. 

 As we are going to have e2e testsfor kubectl, I did not addd multi-tenancy test cases for non-system tenant in this PR.

File: pkg/kubectl/cmd/clusterinfo/clusterinfo.go (hunk 2 above)
Comment (chenqianfzh) [reply pkg/kubectl/cmd/clusterinfo/clusterinfo.go:149]: Updated using a fmt.Sprintf().

File: pkg/kubectl/cmd/cp/cp.go (hunk 3 above)
Comment (chenqianfzh) [reply pkg/kubectl/cmd/cp/cp.go:143]: I think we can just use "3" here, as the code before me is using "1" and "2". The code is quite straightforward and this style is throughout the code base of kubernetes.

**Pull request merged**